realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -633,47 +540,51 @@ protected static synchronized void validateAgainstExistingConfigurations(RealmCo
     /**
      * Deletes the Realm file defined by the given configuration.
      */
-    protected static synchronized boolean deleteRealm(RealmConfiguration configuration) {
-        if (isFileOpen(configuration)) {
-            throw new IllegalStateException("It's not allowed to delete the file associated with an open Realm. " +
-                    "Remember to close() all the instances of the Realm before deleting its file.");
-        }
+    protected static boolean deleteRealm(RealmConfiguration configuration) {
+        synchronized (RealmCache.class) {
```

Comment:
I was really hoping we could avoid something like this, as these kind of rules is what caused the mess in the first place. The only reason for this, seems to be because `isFileOpen` is calling down into the RealmCache?

On way to avoid having the lock here would be to provide a callback class to RealmCache.

RealmCache.isFileOpen(configuration, new Callback() {
  // Callback called from inside RealmCache when locking is guarenteed
});

What do you think?